### DISCUSSION TITLE
maximum poll over ethernet

### DISCUSSION THREAD

**Discussion by EmbeddedDevver** [eModbus/eModbus#168] [Q&A] [answered]
Short question, I just tried eModbus with esp32 over ethernet and WiFi. Working fine. However, I see that I cannot get more than ~1 request each second. In the debugger on my windows client, I see that the esp32 gives a response after 800mS. Is that expected result? Was wondering how to increase the poll per second to get an update around 10 per seconds.

**Comment by Miq1** [comment]
There was a bug causing unnecessary delays of about 100ms per request, but that has been fixed. There is no reason AFAIK why the request should be delayed except for Modbus  timeouts occurring.
Do you have some log files available? We could try to see where the time is spent from a verbose log.

**Comment by EmbeddedDevver** [comment]
Just updated the eModbus library to the current version in this repro. I did a diff with the library that I used yesterday, and not much of a change to be found between the updated library and my used one from yesterday. So I wont expect any difference with the updated / current library.

My setup:
Option A) 
PC to Ethernet
Ethernet to ESP32 with eModbus running in tcp mode.

Option B)
PC to Ethernet
Ethernet to Moxa Modus gateway rs485
Rs485 to Esp32 with eModbus running in RS485 mode

I just did a redo of the tests. Do note, that I also have used an Moxa Modbus RTU gateway in the very same setup (option B). The Moxa can go very fast in like 10 requests per seconds or more. Then replacing the Moxa with eModbus still works (Option A), but at a maximum rate of 1 message per second.

This is my windows application polling the eModbus library for data. 
You can see that the response comes in around 900mS later (yellow is transmit from PC to eModbus).
While the exact same setup using Moxa Gateway (option B) response in less than 70mS.
![image](https://user-images.githubusercontent.com/4362015/160079843-f6efe2c6-2fb6-4725-93ee-02f0a83e154e.png)

Below the output of eModbus during the polls. I think we must look into the ``` handleOutbox();```
```
[D] 165739| ModbusServerTCPasync.cpp [ 165] handleOutbox: sending (29)
[D] 165757| ModbusServerTCPasync.cpp [  35] onData: data len 12
[D] 165757| ModbusServerTCPasync.cpp [  78] onData: request complete (len:12)
[D] 165759| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 165765| ModbusServerTCPasync.cpp [ 115] onData: Data response
[D] 166239| ModbusServerTCPasync.cpp [ 165] handleOutbox: sending (29)
[D] 166244| ModbusServerTCPasync.cpp [  35] onData: data len 12
[D] 166245| ModbusServerTCPasync.cpp [  78] onData: request complete (len:12)
[D] 166247| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 166254| ModbusServerTCPasync.cpp [ 115] onData: Data response
[D] 166739| ModbusServerTCPasync.cpp [ 165] handleOutbox: sending (29)
[D] 166753| ModbusServerTCPasync.cpp [  35] onData: data len 12
[D] 166753| ModbusServerTCPasync.cpp [  78] onData: request complete (len:12)
[D] 166754| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 166761| ModbusServerTCPasync.cpp [ 115] onData: Data response
[D] 167239| ModbusServerTCPasync.cpp [ 165] handleOutbox: sending (29)
[D] 167256| ModbusServerTCPasync.cpp [  35] onData: data len 12
[D] 167257| ModbusServerTCPasync.cpp [  78] onData: request complete (len:12)
[D] 167258| ModbusServer.cpp     [  30] getWorker: Worker found for 01/03
[D] 167264| ModbusServerTCPasync.cpp [ 115] onData: Data response
``` 

Some other interesting item:
When I use a laptop that is connected over WiFi, and then polling the eModbus using the same application, I receive every 2 second a timeout in the windows application saying that remote modbus is not responding / timeout and sometimes even invalid frame received. This does not happen when using the Moxa gateway. So it must be possible to work well over WiFI.
When I connect the laptop using ethernet, than no errors come up, only slow update rate as mentioned above.

Here you can see the response time from the Moxa Modbus Gateway (option B). The response comes back in less than 70mS from the Moxa Modbus gataway.
![image](https://user-images.githubusercontent.com/4362015/160081601-a1256387-b0e2-4abb-ab67-5a3d5844bed1.png)

**Reply by Miq1** [reply]
This is a different story now 😉 

You seem to be using ``ModbusServerTCPasync``, which is much different from ``ModbusServerEthernet`` or ``ModbusServerWiFi`` and is using  completely different lower layers for TCP. As far as I know, the async variant is running on WiFi only, by the way. That is more @bertmelis domain, though, so he better should have a look.

So if you really are using Ethernet, you should use the Ethernet server also.

Another user and me were trying to find the maximum poll rates a few weeks ago and found the minimum time between requests being below 10ms.

**Comment by EmbeddedDevver** [comment, marked as answer]
Resolved here: https://github.com/eModbus/eModbus/pull/169

Yes indeed I am using ```ModbusServerTCPasync```. I am using the native ESP32 library where WiFi and Ethernet can work transparent. For the Modbus library, there is no need to setup something different for WiFi or Ethernet. I just need to connect one of the two interfaces to our network (or both if needed) and then there is directly communication to eModbus. Works great!!

This above option also reduces the amount of external components needed for ethernet communication so it is also a lot cheaper compared to external ethernet chips like wizznet.

Good news; I found the problem and solution to the slow responses. Now I am sitting at 35 responses a second without any error.  And I can monitor that responses now come in at less than 35 milli seconds, awesome!!!

![image](https://user-images.githubusercontent.com/4362015/160100465-7998e97c-df9b-4b06-b6be-7b6b5cbcdc65.png)

I will upload the code in few minutes so you can verify and merge the code.

EDIT
it's even faster when logging is disabled and only 1 client is requesting data, amazing 4mS response time!
![image](https://user-images.githubusercontent.com/4362015/160105249-c1ca4042-4082-43c0-8794-00fd4804bd35.png)
